Revamp server-side callback API and message allocator API

vjpai opened the pull request:
This is a significant API change and I am requesting many sets of eyes on this.

1. Introduce Reactor model for server-side unary calls that allows for OnCancel and separate sending of initial metadata
    - Motivation: `SetCancelCallback`/`ClearCancelCallback` is fragile and hard-to-use (the callback has to be non-blocking, can't perform certain operations, etc). Experience with Stubby internally suggests that these APIs will not actually be used widely if they're the only way of alerting on cancellation. OnCancel from the streaming reactors was already approved and is flexible enough to be useful. Additionally, a small number of applications want to separately get the initial metadata on a unary RPC before getting the full response (possibly to make a decision based on the metadata, etc), so moving to a reactor model allows us to fulfill this use case as well.
1. Change method signature of all server-side callback API RPCs .  This is done by differentiating between `ServerContext` and `CallbackServerContext` as the first parameter for either sync or callback method handlers, respectively.
    - Motivation: Splitting the server context into two variants allowed for a clear overloadable disambiguation between the sync and callback method handlers. This could not have been done using only a different return value since you can't overload on the return value. Other options considered were to collect the reactor in an output parameter (which is not preferred by the Google C++ style guide and is also confusing to read and use) or to add a new unused argument to the head of the argument list (like `enum ::grpc::AsyncMode`, similar to the unused differentiation argument types used in the constructors for `::grpc::Slice`, but this is confusing for providing an argument to a piece of user-written code without any value to interpret) but this solution was considered the best option. It also allows specialization of the two server context types so that they don't include methods only relevant to the other. There is now a common base class to the two server context types, and the server-side interception functions use the common base class.
1. Introduce a new `DefaultReactor` method to the `CallbackServerContext` to allow for responding with a simple pre-allocated reactor that has the functionality previous held by the ServerCallbackRpcController.
    - Motivation: We should still provide a fast and easy-to-use mechanism for minimal or otherwise small unary RPCs that just want to complete and provide their status and response. This will not need a thread-dispatch for OnDone since it has no significant reactions.
1. Introduce a `DefaultReactorTestPeer` to monitor actions taken on the `DefaultReactor`
    - Motivation: To subsume some previous guidance on mocking the `ServerCallbackRpcController` and provide a surface for unit-testing unary methods that use the default reactor.
1. Remove the `OnStarted` call from streaming RPCs.
    - Motivation: The information that was available there is now available instead at the actual method handler itself. Having this information at the method handler time itself gives the user many possible opportunities, like using features of the `CallbackServerContext` or request to make decisions about what type of reactor to create (e.g., if the `CallbackServerContext` is already canceled, just return a finish-only reactor with cancellation rather than doing additional work.) Of note, grpc-node (which in many ways inspires the C++ reactor model) doesn't have a separate split start like the previous C++ reactor model either.
1. Migrate the message allocator API for unary RPCs to the `CallbackServerContext` since that's the piece that is available throughout the RPC lifetime.
    - Motivation: This has to sit in an input parameter for the method handler, and this is the only input parameter that isn't a data message itself.
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/18856)
<!-- Reviewable:end -->

vjpai commented:
___
*[test/cpp/end2end/hybrid_end2end_test.cc, line 812 at r1](https://reviewable.io/reviews/grpc/grpc/18856#-Lf_e04FAKdDPEAWvt_R:-LfcntWqCO7Pe_8AMu6J:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/668233a4cdf2b0a29259b6833ae7066556537eee/test/cpp/end2end/hybrid_end2end_test.cc#L812)):*
<details><summary><i>Previously, vjpai (Vijay Pai) wrote…</i></summary><blockquote>

Saved a variable, so totally worth it.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sheenaqotj commented on include/grpcpp/impl/codegen/server_callback.h:
```diff
@@ -428,16 +391,85 @@ class ServerWriteReactor : public internal::ServerReactor {
+};
+
+template <typename Request, typename Response, typename F>
+ServerUnaryReactor<Request, Response>* ServeRpc(ServerContext* context,
```
I prefer MakeReactor like Sanjay suggested

sheenaqotj replied:
Or maybe MakeServerUnaryReactor

vjpai replied:
Made it MakeReactor.

vjpai commented:
So OnStarted is still the point at which we're telling the reactor that it's ok to start performing operations on the reactor (StartRead, StartWrite, etc). The reactor creator provides a reactor to the library but it's the library's job to decide when it's safe to actually start performing operations on that reactor. Of course, this could potentially be done a different way, so I guess I'll need to look at this some more.

vjpai commented on include/grpcpp/impl/codegen/server_callback.h:
```diff
@@ -435,23 +429,25 @@ class ServerUnaryReactor : public internal::ServerReactor {
-                   Response* resp) override {
-      on_started_func_(context, req, resp, this);
+    void OnStarted(const Request* req, Response* resp) override {
+      on_started_func_(req, resp, this);
```
The user still has to call Finish to indicate done-ness. We could have a "super-simple-reactor" where the user returns a status and the library calls Finish; it could even be an overload...

vjpai replied:
I guess it'll be a different function name rather than an overload since I think the right type_trait to separate these function types is is_invocable which is not valid until C++17 (which we have not yet committed to use)

vjpai replied:
Never mind, enable_if + decltype to the rescue. We can do this as an overload.

vjpai commented on include/grpcpp/impl/codegen/server_callback.h:
```diff
@@ -428,16 +391,85 @@ class ServerWriteReactor : public internal::ServerReactor {
+template <typename Request, typename Response, typename F>
+ServerUnaryReactor<Request, Response>* ServeRpc(ServerContext* context,
+                                                F&& func) {
+  // TODO(vjpai): Specialize this to prevent counting OnCancel conditions
```
Right now we're counting 3 callbacks even though the CompletionOp in this case doesn't ever trigger an OnCancel. But maybe we can get around this in general also with restructuring of OnStarted.

vjpai commented on include/grpcpp/impl/codegen/server_callback.h:
```diff
@@ -998,9 +996,9 @@ class CallbackServerStreamingHandler : public MethodHandler {
+  explicit CallbackBidiHandler(
+      std::function<experimental::ServerBidiReactor<RequestType, ResponseType>*(
+          ServerContext*)>
           func)
```
Great idea; I though that function func is redundant so made it just get_reactor.

vjpai commented on include/grpcpp/impl/codegen/server_callback.h:
```diff
@@ -428,16 +391,85 @@ class ServerWriteReactor : public internal::ServerReactor {
+  ServerCallbackUnary* call_;
+};
+
+template <typename Request, typename Response, typename F>
```
Thanks for the suggestions, I changed it to Function and added a comment.

vjpai commented:
We can eliminate OnStarted by queueing operations started in the reactor getter until the library regains control, similar to the StartCall concept on the client side. This is probably actually better since it also prevents concurrent OnRead/OnWrite with OnStarted

vjpai commented:
AFAIK, I have addressed all existing comments and I think tests will pass. Waiting for that to actually happen and then I'll request reviews.

vjpai commented:
I am still trying to clean up some downstream (internal) breakages, but I believe that this is ready to be reviewed. I will not merge it until the downstream issues are resolved.

vjpai commented:
@sheenaqotj @yang-g can I get reviews on this? I've incorporated all your previous feedback.

vjpai commented:
@yashykt - I'm asking for a primary review from you on the entire PR . Let me know if that's not feasible.

@yang-g - Thanks for your earlier comments. Please see the updates to `server_context_impl.h` and `server_context.cc` as well as the contents of `test/` for topics like `ReactWithStatus` and message allocators.

@mhaidrygoog - Please review the contents of `test/cpp/qps`

yashykt commented:
Thanks Vijay!
1)The motivation for the unary reactor makes sense to me, and it also brings the naming in line with the other RPC types.
2)ReactWithStatus - I'm still not sure about it since it is making additions to server context. Feels like if we provide multiple ways to do the same thing, we will run into bugs.
3) I need to read up on the message allocator API before commenting.
4) Seems fine to me.
5) Where is the need for overloading this coming from?

vjpai commented:
Re ReactWithStatus, the main compelling need is to have a way of supporting unary RPCs that only ever have one user callback since each of those causes a thread dispatch. But I do like your idea, we can do this in a different and more consistent way. Let me pull up a commit on that as soon as I get a laptop....

vjpai commented:
5. The overloading above is required because all variants of the service (sync, async, callback) are derived from the same base Service class (created by the code generator). Although these are invoked by the library, the user has to write the code for the sync or callback versions, so they need to have a function name for writing the method handler. We're not going down the road of adding prefixes or suffixes to the name for callback since those can lead to conflicts with names of user-defined methods in the protos (e.g., if we choose a prefix of "Callback" but the user has a method defined with a name starting with "Callback").

vjpai commented:
The PR got much bigger suddenly, but that's mostly because of moving a lot of content out of `server_callback_impl.h` and into a new file called `server_callback_handlers.h` . The reason is header circular dependences. Previously, `server_callback` depended on `server_context` to supply fields to the method handlers. Now there has to be a header dependence the other way around since a `ServerContext` has a `ServerUnaryReactor` as a member. However, the reactors don't depend on the implementation of `ServerContext` - only yhe method handlers do. So, splitting them out into a separate header allows for that header to depend on both `server_context.h` and `server_callback.h` .

vjpai commented:
@sheenaqotj and @wcevans - can I get your opinions on commit 0d7695ee . Does this provide the scope for performance improvement that you are seeking?

vjpai commented:
___
*[test/cpp/end2end/hybrid_end2end_test.cc, line 818 at r14](https://reviewable.io/reviews/grpc/grpc/18856#-LrPwUP09geMwAHPFU24:-LrR3eVQBaBiU_2K5ft6:b-pb5eix) ([raw file](https://github.com/grpc/grpc/blob/ae847cc6fc83afcfd830050ab5de1d8f6d3210a4/test/cpp/end2end/hybrid_end2end_test.cc#L818)):*
<details><summary><i>Previously, mhaidrygoog (Moiz Haidry) wrote…</i></summary><blockquote>

Would it make sense to then have a pointer to the method and not the ctx?
</blockquote></details>

Yes! Great suggestion. We can do even better by moving this check into the method handler and just using the reactor constructor for StartRead.


<!-- Sent from Reviewable.io -->

vjpai commented on include/grpcpp/impl/codegen/server_context_impl.h:
```diff
@@ -272,6 +279,22 @@ class ServerContext {
+
+  /// Get the default unary reactor for use in minimal reaction cases.
+  /// WARNING: This is experimental API and could be changed or removed.
+  experimental::ServerUnaryReactor* DefaultReactor() {
```
I'd like to ban it so that we can do something in this case later (like lazy initialization as Alan suggested).

vjpai replied:
But I don't know if banning it is the right decision.

vjpai replied:
So at the current time, it's ok to call it multiply. Maybe we can decide if it should be that way.

hcaseyal replied:
What should a user expect? If a user calls this twice, are they going to think that they got two fresh reactors? Are they going to think they need to free each one, especially since the typical non-DefaultReactor() use case has users deeply allocating the reactors? I'm concerned that users will get very confused by this.

Alternatively, we could provide a factory method that constructs a default reactor.

vjpai replied:
Added comments to make the calling convention more clear and explicitly forbid multiple calls.

vjpai commented:
*[a discussion](https://reviewable.io/reviews/grpc/grpc/18856#-Lfa2HCeBgMK06syyKFB:-LrjJ-XbCsOfpAzXclhk:bltxx46) (no related file):*
<details><summary><i>Previously, vjpai (Vijay Pai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This comment chain is based on an obsolete version of the PR and is now resolved with the removal of OnStarted.

---


<!-- Sent from Reviewable.io -->

hcaseyal commented on include/grpcpp/impl/codegen/server_context_impl.h:
```diff
@@ -272,6 +279,36 @@ class ServerContext {
+    return message_allocator_state_;
+  }
+
+  /// Get the default unary reactor for use in minimal reaction cases.
```
We should make it clear that the user doesn't need to free or delete the returned reactor

vjpai replied:
I added the words "library-owned" as well as a more detailed description to this comment.

vjpai replied:
In reply to your other comment (which I somehow can't reply to through the UI), we're not using a factory precisely because this is supposed to avoid an allocation. I added a comment saying that they shouldn't call this more than once; although there's no enforcement mechanism now or even any negative impact, we may take advantage of this restriction in the future (e.g., to ease lazy initialization of the DefaultReactor).

vjpai commented:
This now addresses all the design and internal review comments. Please consider it reviewable again. Thanks!

vjpai commented:
@hcaseyal I think I've addressed your concerns on the `DefaultReactor` docs and the `DefaultReactorTestPeer` . Can I get your re-reviews on those parts?

soheilhy commented on include/grpcpp/impl/codegen/server_callback_impl.h:
```diff
@@ -222,34 +215,55 @@ class ServerCallbackReaderWriter {
   /// Finish (if there are no writes).
-  void StartSendInitialMetadata() { stream_->SendInitialMetadata(); }
+  void StartSendInitialMetadata() {
+    auto* stream = stream_.load(std::memory_order_acquire);
```
nit: as per c-style, please spell out auto type because the type is not obvious within +/-10 lines

soheilhy commented on include/grpcpp/impl/codegen/server_callback_impl.h:
```diff
@@ -222,34 +215,55 @@ class ServerCallbackReaderWriter {
+    auto* stream = stream_.load(std::memory_order_acquire);
+    if (stream == nullptr) {
+      grpc::internal::MutexLock l(&stream_mu_);
+      stream = stream_.load(std::memory_order_acquire);
```
The second load for double-checked locking should be `relaxed` because it's synchronized with mutex.  The first loads/stores should be acquire/release pairs.

soheilhy commented on include/grpcpp/impl/codegen/server_callback_impl.h:
```diff
@@ -222,34 +215,55 @@ class ServerCallbackReaderWriter {
+    auto* stream = stream_.load(std::memory_order_acquire);
+    if (stream == nullptr) {
+      grpc::internal::MutexLock l(&stream_mu_);
+      stream = stream_.load(std::memory_order_acquire);
```
this can similarly be `relaxed` here and below

soheilhy commented on include/grpcpp/impl/codegen/server_callback_impl.h:
```diff
@@ -353,835 +393,308 @@ class ServerBidiReactor : public internal::ServerReactor {
+    }
+    if (finish_wanted_) {
+      reader->Finish(std::move(status_wanted_));
+    }
```
Is it worth resetting these fields after being processed once?

For example, if `read_wanted_`  is not `nullptr`, we should run Read and also set `read_wanted_` to `nullptr`?

(same question for the writer)

vjpai replied:
Hypothetically (and it's hypothetical because we can't use C++17 right now here), could we use a std::variant between the `reader` and a struct containing the `_wanted`? I think that would give the strongest statement that the combination of these things is not relevant at the same time (though I'm not sure how `variant` plays with `atomic`). That said, we can reset values to nullptr to ease readability, I'm fine with that.

soheilhy replied:
Thank you for add the resets.  It just occurred to me to ask whether it's ok to call `Finish()`, `Read()` and `SendInitialMetadata()` while holding the lock? My question is mainly about performance.

We can potentially just get and reset the pointers and call the stream methods after Lock is released.

vjpai replied:
Well, that's interesting. It's safe from a lock-inversion pov etc but yeah, there may be a performance impact especially for streaming cases. I'll take a crack at this right away.

vjpai commented:
@soheilhy - although I'd like to pick up what you requested, can we do that performance improvement in a follow-on PR?

soheilhy commented:
@vjpai yes, certainly! thanks.

hcaseyal commented on test/cpp/end2end/mock_test.cc:
```diff
@@ -161,6 +162,86 @@ class FakeClient {
+
+  req.set_message("ha ha, consider yourself mocked.");
+  auto* reactor = service_.Echo(&ctx, &req, &resp);
+  EXPECT_EQ(reactor, peer.reactor());
```
If the reactor we get from the service is the same as the reactor we get from the peer, what's the point in exposing it from the peer?

vjpai replied:
This is a check that the service is actually using the default reactor. If the service isn't using the default reactor, the DefaultReactorTestPeer won't be able to be used.

hcaseyal replied:
It would be better if this was automatically checked so that it fails if the user tries to mock a call to a service that doesn't use the default reactor. In the current state, if the user's service doesn't use the default reactor and they forget to add this EXPECT_EQ to their test, they might waste cycles debugging mysterious behavior. 

Since every call to DefaultReactorTestPeer is going to be followed by an invocation of a method on the service, why don't we let the DefaultReactorTestPeer invoke the user's method? Then it can always perform the pointer check. Something like:

DefaultReactorTestPeer peer(&ctx, &service_.Echo);
peer.invoke_method() // Invokes Echo and does the EXPECT_EQ(reactor, peer.reactor()) check

hcaseyal replied:
Something like:

DefaultReactorTestPeer peer(&ctx);
peer.invoke_method(&service_.Echo, &req, &resp) // Invokes Echo and does the EXPECT_EQ(reactor, peer.reactor()) check

vjpai replied:
I've filed #21228 to make sure that we do something about this, preferably as additive API.

vjpai commented on include/grpcpp/impl/codegen/server_context_impl.h:
```diff
@@ -88,36 +95,29 @@ namespace grpc {
 namespace testing {
 class InteropServerContextInspector;
 class ServerContextTestSpouse;
+class DefaultReactorTestPeer;
```
https://abseil.io/tips/135 consistently calls it a Peer and says that it's sometimes called a spouse. (This was earlier in an internal ToTW 135).  We can't change the name of the Spouse now (to avoid breaking existing users) but we shouldn't add deprecated terms.

hcaseyal replied:
I don't see anything in there that suggests Spouse is deprecated.

It would be nice if we could be internally consistent, but I honestly prefer the term "peer" over "spouse", so I guess this is fine.

vjpai commented on test/cpp/end2end/mock_test.cc:
```diff
@@ -161,6 +162,86 @@ class FakeClient {
+class MockCallbackTest : public ::testing::Test {
+ protected:
+  CallbackTestServiceImpl service_;
+
```
Done.

vjpai commented on test/cpp/end2end/mock_test.cc:
```diff
@@ -161,6 +162,86 @@ class FakeClient {
+  experimental::ServerUnaryReactor* Echo(
+      experimental::CallbackServerContext* context, const EchoRequest* request,
+      EchoResponse* response) override {
+    // Make the mock service return a choice of Status even though the
```
I have expanded this comment now.

vjpai commented on include/grpcpp/test/default_reactor_test_peer.h:
```diff
@@ -0,0 +1,50 @@
+namespace grpc {
+namespace testing {
+
+/// A test-only class to access private members and methods of ServerContext.
```
I have expanded this comment. I'm a little loath to add more examples here because that will be covered separately in the codelab and also because we don't encourage mocking (it's a necessary evil for our broader codebase though).

vjpai commented on include/grpcpp/impl/codegen/server_context_impl.h:
```diff
@@ -268,13 +273,40 @@ class ServerContext {
+    return message_allocator_state_;
+  }
+
+  /// Get a library-owned default unary reactor for use in minimal reaction
```
Added, along with explaining the use of optimized minimal reactions.

vjpai commented:
@soheilhy - though I did realize that the actual "finish" action shouldn't be invoked under the lock as part of the BindCall/etc functions, since finish can end up triggering ondone which can end up destroying the reactor object including its still-held lock.

soheilhy commented:
@vjpai thank you for the note! LGTM.

hcaseyal commented on include/grpcpp/impl/codegen/server_context_impl.h:
```diff
@@ -268,13 +273,44 @@ class ServerContext {
+  /// or early sending of initial metadata. Since this is a library-owned
+  /// reactor, it should not be delete'd or freed in any way. This is more
+  /// efficient than creating a user-owned reactor both because of avoiding an
+  /// allocation and because its minimal reactions are optimized using a core
```
It's worth being clearer about exactly what allocation we're avoiding. Something like:
"DefaultReactor() is more efficient than creating a user-owned reactor because it avoids a separate allocation for the reactor. In addition, the default reactor's reactions are optimized using..."

yashykt commented on include/grpc/impl/codegen/grpc_types.h:
```diff
@@ -726,6 +726,10 @@ typedef struct grpc_experimental_completion_queue_functor {
 
+  /** The inlineable member specifies whether this functor can be run inline.
+      This should only be used for trivial internally-defined functors. */
+  int inlineable;
```
non-blocking comment - Do we foresee this field being used by users, i.e., will we need to provide a knob to users to set this?

vjpai replied:
Addressed on reviewable

yashykt commented on include/grpcpp/impl/codegen/callback_common.h:
```diff
@@ -85,6 +85,10 @@ class CallbackWithStatusTag
       : call_(call), func_(std::move(f)), ops_(ops) {
     g_core_codegen_interface->grpc_call_ref(call);
     functor_run = &CallbackWithStatusTag::StaticRun;
+    // A client-side callback should never be run inline since they will always
```
non-blocking comment - never sounds like a strong word.. what if it's a fast no IO function?

yashykt commented on include/grpcpp/impl/codegen/server_callback_handlers.h:
```diff
@@ -0,0 +1,797 @@
+    /// the completion of the RPC. This should be the last component of the
+    /// handler.
+    void SetupReactor(experimental::ServerUnaryReactor* reactor) {
+      reactor_.store(reactor, std::memory_order_relaxed);
```
Why is relaxed memory ordering fine here? Please add some documentation

vjpai replied:
Done

yashykt commented on include/grpcpp/impl/codegen/server_callback_impl.h:
```diff
@@ -353,835 +398,355 @@ class ServerBidiReactor : public internal::ServerReactor {
   }
 
-  ServerCallbackReaderWriter<Request, Response>* stream_;
+  grpc::internal::Mutex stream_mu_;
```
let's add notes on which fields require locking and which do not

vjpai replied:
Done

yashykt commented on include/grpcpp/impl/codegen/server_callback_impl.h:
```diff
@@ -353,835 +398,355 @@ class ServerBidiReactor : public internal::ServerReactor {
 
-  ServerCallbackReaderWriter<Request, Response>* stream_;
+  grpc::internal::Mutex stream_mu_;
+  std::atomic<ServerCallbackReaderWriter<Request, Response>*> stream_;
```
is the reason for stream_ to be accessed without a lock and then again inside a lock, simply a Test and Test and Set kind of optimization (more like Load and Locked Load)?

vjpai replied:
Yup!

vjpai commented:
All needed LGTMs are in. I'm going to rebase...
